src/lib/interlink.lua

```diff
@@ -0,0 +1,245 @@
+--    pull(r)                    push(r)
+--    detach_receiver(name)      detach_transmitter(name)
+--
+-- I.e., both receiver and transmitter will attach to a queue object they wish
+-- to communicate over, and detach once they cease operations.
+--
+-- Meanwhile, the receiver can extract packets from the queue unless it is
+-- empty, while the transmitter can insert new packets into the queue unless
+-- it is full.
+--
+-- Packets inserted by the transmitter only become visible to the receiver once
+-- the transmitter calls push. Likewise, queue slots freed from extracting
+-- packets only become visible to the transmitter once the receiver calls pull.
+--
+-- API
+-- ----
+--
+--    attach_receiver(name), attach_transmitter(name)
+--       Attaches to and returns a shared memory interlink object by name (a
+--       SHM path). If the target name is unavailable (possibly because it is
+--       already in use), this operation will block until it becomes available
+--       again.
+--
+--    detach_receiver(r, name), detach_transmitter(r, name)
+--       Unmaps interlink r and unlinks it from its name. If other end has
+--       already freed the interlink, any packets remaining in the queue are
+--       freed.
+--
+--    full(r) / empty(r)
+--       Return true if the interlink r is full / empty.
+--
+--    insert(r, p) / extract(r)
+--       Insert a packet p into / extract a packet from interlink r. Must not
+--       be called if r is full / empty.
+--
+--    push(r) / pull(r)
+--       Makes subsequent calls to full / empty reflect updates to the queue
+--       caused by insert / extract.
+
+local shm = require("core.shm")
+local ffi = require("ffi")
+local band = require("bit").band
+local waitfor = require("core.lib").waitfor
+local sync = require("core.sync")
+
+local SIZE = link.max + 1
+local CACHELINE = 64 -- XXX - make dynamic
+local INT = ffi.sizeof("int")
+
+assert(band(SIZE, SIZE-1) == 0, "SIZE is not a power of two")
+
+-- Based on MCRingBuffer, see
+--   http://www.cse.cuhk.edu.hk/%7Epclee/www/pubs/ipdps10.pdf
+
+ffi.cdef([[ struct interlink {
+   int read, write, state[1];
+   char pad1[]]..CACHELINE-3*INT..[[];
+   int lwrite, nread;
+   char pad2[]]..CACHELINE-2*INT..[[];
+   int lread, nwrite;
+   char pad3[]]..CACHELINE-2*INT..[[];
+   struct packet *packets[]]..SIZE..[[];
+} __attribute__((packed, aligned(]]..CACHELINE..[[)))]])
+
+-- The life cycle of an interlink is managed using a state machine. This is
+-- necessary because we allow receiving and transmitting processes to attach
+-- and detach in any order, and even for multiple processes to attempt to
+-- attach to the same interlink at the same time.
+--
+-- Interlinks can be in one of five states:
+
+local FREE = 0 -- Implicit initial state due to 0 value.
+local RXUP = 1 -- Receiver has attached.
+local TXUP = 2 -- Transmitter has attached.
+local DXUP = 3 -- Both ends have attached.
+local DOWN = 4 -- Either end has detached; must be re-allocated.
```

Comment:
I don't quite get this description; the state transition diagram below suggests that DXUP -> TXUP is valid.  I would have assumed that DXUP can only transition to DOWN, based on the description.  What am I missing?

Reply:
Should the comment be, "both ends have detached" ?

Reply:
Yep that’s right the comments were off.  :-/  An interlink can go from DXUP to {T,R}XUP back to DXUP. It must deallocated once both ends have detached. I.e., more than two processes can attach to a single interlink during its life time. As long as either end is attached, both receiver and transmitter can be superseded by new processes. Should be fixed now.